Warn when a supplied mapping is going to be overwritten by geom_hline / vline / abline

RichardJActon opened the pull request:
This pull request addressed issue #2945

geom_hline / vline / abline will now throw a warning if the user supplies both a x|y|intercept/slope value and a mapping such as this:

    geom_hline(yintercept = 3, aes(colour = colour))

As the above will cause the user supplied mapping to be ignored/overwritten

Users might resaonably expect the above to work as other geoms do not have this behaviour. The warning messages suggest that the x|y|intercept/slope value(s) be moved inside the aes call thus:

    geom_hline(aes(yintercept = 3,colour = colour))

hadley commented on R/geom-hline.r:
```diff
@@ -9,6 +9,15 @@ geom_hline <- function(mapping = NULL, data = NULL,
                        show.legend = NA) {
 
+  # Warn if supplied mapping is going to be overwritten
+  if (!missing(yintercept) && !missing(mapping)) {
```
This would be better placed in the block below

hadley replied:
I think you forgot to remove the code here

yutannihilation commented:
Is it OK to recommend to place values directly in `aes()`? In this case, it's fine, but I'm a bit afraid this may be a bit confusing because using direct values in `aes()` is not always good.

Maybe preparing data is the right thing? Or considering the hacky nature of the "annotation mode", is this allowed specially?

clauswilke commented:
> Maybe preparing data is the right thing?

Actually, `geom_hline()` will overwrite the dataset if one is provided, so this won't work. I've been wondering whether the code block here:
https://github.com/tidyverse/ggplot2/blob/35507725d82d1d9d91fe5d2e4943671b568b59b2/R/geom-hline.r#L13-L17
should only be executed when data is `NULL`.

yutannihilation commented:
I meant not providing `*intercept` and map all the variables.

```r
data2 <- data
data2$y <- 3

geom_hline(data = data2,
           aes(yintercept = y, colour = colour))
```

But, please ignore my comment above. I think I don't get the setantics of variable mappings yet. Sorry for the noise.

hadley commented on R/geom-hline.r:
```diff
@@ -9,8 +9,26 @@ geom_hline <- function(mapping = NULL, data = NULL,
   if (!missing(yintercept)) {
+    # Warn if supplied mapping is going to be overwritten
+    if (!missing(yintercept) && !missing(mapping)) {
+      warning(paste0("Using both `yintercept` and `mapping` may not have the desired result as mapping is overwritten if `yintercept` is specified\n",
```
You don't need to check `!missing(yintercept)` again (this is the reason that this code belongs in this block).

For now, I think it would be better if the warning message simply stated what was happening (i.e. setting `yintercept` overrides `mapping`) rather than recommending a solution.

hadley commented on tests/testthat/test-geom-hline-vline-abline.R:
```diff
@@ -43,3 +43,38 @@ test_that("curved lines in map projections", {
+# Warning tests ------------------------------------------------------------
+
+test_that("Warning if a supplied mapping is going to be overwritten",{
+  dat <- data.frame(x = LETTERS[1:5], y = 1:5,colour = c(TRUE,TRUE,rep(FALSE,3)))
```
There's no need to construct a plot here since the warning comes from the `geom_vline()` call.

RichardJActon commented:
Are any further changes needed here?

- I have placed the warning block in the existing check for a missing intercept value and removed the redundant test for the intercept.
- I removed the plot from the test and just checked the geoms directly
- I altered the error message so it does suggest the solution of placing a value directly in a mapping

(I'm not sure what's up with travis - the error says something about exceeding the API rate limit?)

clauswilke commented on R/geom-abline.r:
```diff
@@ -82,6 +82,16 @@ geom_abline <- function(mapping = NULL, data = NULL,
+
+    # Warn if supplied mapping is going to be overwritten
+    if (!missing(mapping)) {
+      warning(paste0("Using both `intercept` and/or `slope` with `mapping` may",
```
I think the word "both" is misplaced, as it refers to the two words "intercept" and "slope". Maybe write: "Using `intercept` and/or `slope` in combination with `mapping` may"

clauswilke commented on R/geom-abline.r:
```diff
@@ -82,6 +82,16 @@ geom_abline <- function(mapping = NULL, data = NULL,
 
   # Act like an annotation
   if (!missing(slope) || !missing(intercept)) {
```
I still think this code path should only be executed if no dataset is provided. It's not intuitive to overwrite a dataset the user has provided (and here this also happens without any kind of warning or message.) I can't envision any scenario in which replacing an explicitly provided dataset would be the right course of action.

clauswilke commented on tests/testthat/test-geom-hline-vline-abline.R:
```diff
@@ -43,3 +43,33 @@ test_that("curved lines in map projections", {
+  )
+
+  expect_warning(
+    geom_abline(intercept = 3,slope=0.5, aes(colour = colour)),
```
Please add a space after the comma.

hadley replied:
And around `=`

clauswilke commented:
This cannot currently be merged because the code in `geom_abline()`, `geom_hline()`, and `geom_vline()` has changed.

@RichardJActon Would you mind updating this? Could you please also add a bullet to NEWS.md? It should briefly [describe the change](http://style.tidyverse.org/news.html) and end with `(@yourname, #issuenumber)`.

clauswilke commented:
We just opened a separate issue about the overwriting of `data`: #3090 I suggest you finish this PR as it currently stands.

clauswilke commented on tests/testthat/test-geom-hline-vline-abline.R:
```diff
@@ -43,3 +43,33 @@ test_that("curved lines in map projections", {
+
+# Warning tests ------------------------------------------------------------
+
+test_that("Warning if a supplied mapping is going to be overwritten",{
```
Could you please add a space between the comma and the opening curly brace?

clauswilke commented on NEWS.md:
````diff
@@ -1,4 +1,13 @@
+ 
+  ```
+  
+  The above will cause the user supplied mapping to be ignored/overwritten and as users might resaonably expect the above to work given that other geoms do  not have this behaviour, they will now be warned about mapping being overwritten. (@RichardJActon, #2950)
````
This news item seems overly verbose. Something like the first sentence (on line 3) would be sufficient. Also, please write "and" instead of "&" throughout, and please use the Oxford comma: `` `geom_hline()`, `geom_vline()`, and `geom_abline()` ``. Finally, please place the final period after `(@RichardJActon, #2950)`, not before.

clauswilke replied:
Two more comments on the same topic:
1. `xintercept`, `yintercept`, etc should be enclosed in backticks in the text.
2. The Oxford comma also applies to "or".
 
I.e.: `` `xintercept`, `yintercept`, or `slope` ``

clauswilke commented:
Thanks!
